scheduling_group: improve scheduling group creation exception safety

mlitvk opened the pull request:
Improve handling of exceptions during scheduling group and scheduling group key creation, where a user-provided constructor for the keys may fail, for example.

We use a new struct `specific_val` and smart pointers to manage memory allocation, construction and destruction of scheduling group data in a safe manner.

We also reorder the initialization order to make it safer. For example, when creating a scheduling group, first allocate all data and then swap it into the scheduling group's data structure.

Fixes scylladb/seastar#2222

mlitvk commented:
The CI fails in Seastar.unit.rpc with timeout
This is a known issue: https://github.com/scylladb/seastar/issues/2620

xemul commented on include/seastar/core/scheduling_specific.hh:
```diff
@@ -37,17 +37,66 @@ namespace seastar {
 
 struct scheduling_group_specific_thread_local_data {
+    using val_ptr = std::unique_ptr<void, void (*)(void*) noexcept>;
+    using cfg_ptr = std::shared_ptr<scheduling_group_key_config>;
```
Can it be `seastar::lw_shared_ptr<scheduling_group_key_config>`?

mlitvk replied:
done

xemul commented on include/seastar/core/scheduling_specific.hh:
```diff
@@ -37,17 +37,66 @@ namespace seastar {
 namespace internal {
 
 struct scheduling_group_specific_thread_local_data {
+    using val_ptr = std::unique_ptr<void, void (*)(void*) noexcept>;
```
What's the point in making it smart-pointer if you track construction/destruction by hand anyway?

mlitvk replied:
The point is to manage the dynamic memory allocation and free it automatically

xemul commented on include/seastar/core/reactor.hh:
```diff
@@ -322,6 +322,14 @@ private:
+    inline auto& get_sg_data(unsigned sg_id) {
+        return _scheduling_group_specific_data.per_scheduling_group_data[sg_id];
+    }
+
```
There's no need in those helpers, AFAICS, the existing `get_scheduling_group_specific_thread_local_data()` (and Co) already provide access to the array of per_scheduling_group_data-s

mlitvk replied:
I added it because in several places we access the sg data and we do
```
    auto& sg_data = _scheduling_group_specific_data;
    auto& this_sg = sg_data.per_scheduling_group_data[sg._id];
```
which I thought is a little cumbersome and I didn't find another method for this

xemul commented on include/seastar/core/reactor.hh:
```diff
@@ -393,7 +401,6 @@ private:
     task_queue* pop_active_task_queue(sched_clock::time_point now);
     void insert_activating_task_queues();
     void account_runtime(task_queue& tq, sched_clock::duration runtime);
-    void allocate_scheduling_group_specific_data(scheduling_group sg, unsigned long key_id);
```
I'd appreciate if the change "Move allocate_scheduling_group_specific_data() from reactor class to internal namespace" was made as separate preparational patch (in the same PR)

mlitvk replied:
done

xemul commented on include/seastar/core/scheduling_specific.hh:
```diff
@@ -37,17 +37,66 @@ namespace seastar {
+
+        specific_val() : valp(nullptr, &free), cfg(nullptr) {}
+
+        specific_val(val_ptr&& valp_, const cfg_ptr& cfg_) : valp(std::move(valp_)), cfg(cfg_) {
```
I'm probably misreading the intent, but  specific_val::cfg  and scheduling_group_specific_thread_local_data::scheduling_group_key_configs are intended to point to the _same_ shared object. But the way this constructor is defined makes it impossible to _share_ the config, the specific_val gets its _copy_ of it. Please, correct me if I'm wrong.

xemul replied:
Ah, yes, my bad. It's cfg_ptr

mlitvk commented:
* changed to `lw_shared_ptr` instead of `std::shared_ptr`
* split the commit and added a preliminary commit to change `allocate_scheduling_group_specific_data` into an internal function
* rebase

xemul commented on include/seastar/core/scheduling_specific.hh:
```diff
@@ -37,17 +37,66 @@ namespace seastar {
+        specific_val(const specific_val& other) = delete;
+        specific_val& operator=(const specific_val& other) = delete;
+
+        specific_val(specific_val&& other) : valp(std::move(other.valp)), cfg(other.cfg) {}
```
Why not `cfg(std::move(other.cfg))`? In general, the lifetime of `specific_val::cfg` is unclear.

Here `other.cfg` is kept non-nullptr, what for? Presumably (but maybe I'm wrong) this is to make `~specific_val()` avoid checking if `cfg` exists or not? But the default constructor `specific_val()` sets cfg to nullptr, so it looks like `cfg` _can_ be null, so why not std::move() it here as well?

mlitvk replied:
done, I changed to `std::move`
there was no special reason to copy instead of move, I thought it doesn't matter much, but I suppose it's better to move

xemul commented on include/seastar/core/scheduling_specific.hh:
```diff
@@ -37,17 +37,66 @@ namespace seastar {
     };
     std::array<per_scheduling_group, max_scheduling_groups()> per_scheduling_group_data;
-    std::map<unsigned long, scheduling_group_key_config> scheduling_group_key_configs;
+    std::map<unsigned long, cfg_ptr> scheduling_group_key_configs;
```
It looks like this map is no longer needed? The very came config can be obtained via `per_scheduling_group_data[context.sg_id].specific_vals[key_id].cfg`? I'm not proposing to change anything right now, just checking if my understanding is correct

mlitvk replied:
It's still used in `init_scheduling_group` when we init a new scheduling group and go over all the keys to allocate them for the new sg

xemul commented on include/seastar/core/scheduling_specific.hh:
```diff
@@ -37,17 +37,66 @@ namespace seastar {
+        val_ptr valp;
+        cfg_ptr cfg;
+
+        specific_val() : valp(nullptr, &free), cfg(nullptr) {}
```
Isn't `free` defaulted by compiler for unique-ptr?

mlitvk replied:
the default uses `delete` and it doesn't work because `can't delete pointer to incomplete type`

xemul commented on src/core/reactor.cc:
```diff
@@ -4896,81 +4885,81 @@ deallocate_scheduling_group_id(unsigned id) noexcept {
+    using val_ptr = internal::scheduling_group_specific_thread_local_data::val_ptr;
+    using specific_val = internal::scheduling_group_specific_thread_local_data::specific_val;
+
+    val_ptr valp(aligned_alloc(cfg->alignment, cfg->allocation_size), &free);
```
Isn't `free` defaulted by compiler for unique-ptr?

mlitvk replied:
answered above

xemul commented on src/core/reactor.cc:
```diff
@@ -4896,81 +4885,81 @@ deallocate_scheduling_group_id(unsigned id) noexcept {
+
+    val_ptr valp(aligned_alloc(cfg->alignment, cfg->allocation_size), &free);
+    if (!valp) {
+        throw std::runtime_error("memory allocation failed");
```
It was `std::abort()` before this patch. Why did you change that?

mlitvk replied:
because the intention of the patch is to improve error handling and allow the node to continue to function in face of unexpected errors
so if we can handle exceptions in this path I don't think there's reason anymore to abort in this specific case

xemul commented on src/core/reactor.cc:
```diff
@@ -4896,81 +4885,81 @@ deallocate_scheduling_group_id(unsigned id) noexcept {
             }
             return make_ready_future();
+        }).then([this, key_id, cfgp] () {
+            _scheduling_group_specific_data.scheduling_group_key_configs[key_id] = std::move(cfgp);
```
You seem to mention this change in patch comment:

> We also reorder the initialization order to make it safer. For
example, when creating a scheduling group, first allocate all data and
then swap it into the scheduling group's data structure.

but don't explain the motivation. Why can't scheduling_group_key_config[key_id] be assigned where it was before this change?

mlitvk replied:
it is to avoid the case of a partially initialized scheduling group.
if we allocate some of the values, transfer the ownership to the sg, and then fail, then we remain with a partially initialized scheduling which could cause problems, like if someone assumes it's all be initialized, or if these allocations "leak" because the operation is considered to have failed but we still store the data.
so instead we either allocate all keys or none

xemul replied:
The description you gave applies to `init_scheduling_group` -- it iterates over `for (const auto& [key_id, cfg] : sg_data.scheduling_group_key_configs)`, creates temporary `vals` object, then assigns it to sched group data with `std::swap(this_sg.specific_vals, vals);`.

My concern is about `init_new_scheduling_group_key` which looks like this (simplified):
```
    return parallel_for_each (task_queues, [] (tq) {
        sched_groups[tq.id].specific_vals[key] = allocate_sg_specific_data();
    }).then([] {
        sched_group_configs[key] = config;
    });
```

So it may happen that some sched_groups[tq.id] have its specific_vals[key] set and some not, so this

> ... we either allocate all keys or none

seems not to be true. Neither I see the point in assigning config _after_ all values are allocated. Please, elaborate.

mlitvk replied:
ok, I changed it to be like before.

mlitvk commented:
* in move constructor of `specific_val`, changed it to move the cfg shared ptr instead of copy
* rebase

mlitvk commented:
* changed the initialization order of sg and sg key back to be like before
